### PR TITLE
8266742: Check W^X state on possible safepoint

### PR DESCRIPTION
**Pull request by AntonKozlov**
Hi,

Please review a check for W^X mode at safepoint. As described in the bug, this explicit check aims to catch the wrong W^X mode that may lead to a crash (such crashes are rather intermittent).

This check helped to verify the fix for https://bugs.openjdk.java.net/browse/JDK-8265292 and should help to discover similar issues. It is disabled by default to avoid unexpected failures in the regular use, but eventually it should be turned on unconditionally.

I still owe a complete W^X approach description. Hope this patch does not depend much on that.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266742](https://bugs.openjdk.java.net/browse/JDK-8266742): Check W^X state on possible safepoint

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3920/head:pull/3920` \
`$ git checkout pull/3920`

Update a local copy of the PR: \
`$ git checkout pull/3920` \
`$ git pull https://git.openjdk.java.net/jdk pull/3920/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3920`

View PR using the GUI difftool: \
`$ git pr show -t 3920`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3920.diff">https://git.openjdk.java.net/jdk/pull/3920.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/bsd/globals_bsd.hpp [hunk 1]
```diff
@@ -28,13 +28,16 @@
...
-                         constraint)
+#define RUNTIME_OS_FLAGS(develop,                                       \
+                         develop_pd,                                    \
+                         product,                                       \
+                         product_pd,                                    \
+                         notproduct,                                    \
+                         range,                                         \
+                         constraint)                                    \
+                                                                        \
+  AARCH64_ONLY(develop(bool, WXCheckAtSafepoint, false,                 \
```
Comment (dholmes-ora) [inline src/hotspot/os/bsd/globals_bsd.hpp:39]: Why not enable always as this is a develop build? Do we even need to manage this on a flag?

Also not really "safepoint" as the check also applies to handshakes.

File: src/hotspot/share/runtime/thread.hpp [hunk 2]
```diff
@@ -684,6 +684,10 @@ class Thread: public ThreadShadow {
  public:
   void init_wx();
   WXMode enable_wx(WXMode new_state);
+
+  void check_wx(WXMode expected) {
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.hpp:688]: assert_wx_state ?

**Review by dholmes-ora — COMMENTED** [review]
Hi Anton,

A few comments.

Thanks,
David

File: src/hotspot/os/bsd/globals_bsd.hpp (hunk 1 above)
Comment (AntonKozlov) [reply src/hotspot/os/bsd/globals_bsd.hpp:39]: Even in debug builds the check may be to assertive. This checks exposes 22 failures in tier1 in jfr, gc, compiler, serviceability, and foreign (hope to be fixed soon), and they look like actual possible issues. So I suggest to have the option but keep it disabled.

**Comment by AntonKozlov** [comment]
Thanks! I've changed the option and the method name.

File: src/hotspot/share/runtime/safepointMechanism.inline.hpp [hunk 3]
```diff
@@ -65,6 +65,16 @@ bool SafepointMechanism::should_process(JavaThread* thread) {
 }
 
 void SafepointMechanism::process_if_requested(JavaThread* thread) {
```
Comment (gerard-ziemski) [inline src/hotspot/share/runtime/safepointMechanism.inline.hpp:67]: Looking at the stack from the crash:

```
V [libjvm.dylib+0x8c1060] _ZN21MarkActivationClosure12do_code_blobEP8CodeBlob+0x3c 
V [libjvm.dylib+0x8fea78] _ZN10JavaThread11nmethods_doEP15CodeBlobClosure+0x84 
V [libjvm.dylib+0x401100] _ZN14HandshakeState18process_self_innerEv+0x18c 
V [libjvm.dylib+0x400f60] _ZN14HandshakeState15process_by_selfEv+0x48 
V [libjvm.dylib+0x832690] _ZN18SafepointMechanism25process_if_requested_slowEP10JavaThread+0x54 
V [libjvm.dylib+0x8fe4b8] _ZN10JavaThread44check_safepoint_and_suspend_for_native_transEPS_+0x24 
V [libjvm.dylib+0x92395c] _ZN19ProgrammableInvoker13invoke_nativeEPFvPhES0_P10JavaThread+0x118 
V [libjvm.dylib+0x923aa8] PI_invokeNative+0x104 
J 2565 jdk.internal.foreign.abi.ProgrammableInvoker.invokeNative(JJ)V jdk.incubator.foreign@17-ea (0 bytes) @ 0x00000001189eeed0 
```

why aren't we adding the assertion closer to the frame that crashed, where we need it, like for example:

`HandshakeState::process_self_inner()`

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/safepointMechanism.inline.hpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/safepointMechanism.inline.hpp:67]: If we push the check too low we only do it when we hit this code and there is something to process, but we want to detect when we reach this code in the wrong state even if we wouldn't actually crash at this time due to there not being an active poll.

David

**Review by dholmes-ora — APPROVED** [review]
Changes looks good to me.

Thanks,
David

File: src/hotspot/share/runtime/safepointMechanism.inline.hpp (hunk 3 above)
Comment (AntonKozlov) [reply src/hotspot/share/runtime/safepointMechanism.inline.hpp:67]: David, thanks, this is exactly how it meant to be.

The crash happens when a thread is synchronized for MarkActivationClosure. Having the check in the HandshakeState::process_self_inner would catch the wrong state when a thread is synchronized also for something else. The patch catches the wrong state if a thread polls synchronization, it does not need to actually synchronize to trigger the check. So we'll catch more.

**Review by gerard-ziemski — APPROVED** [review]
Looks good, thanks!

File: src/hotspot/share/runtime/safepointMechanism.inline.hpp (hunk 3 above)
Comment (gerard-ziemski) [reply src/hotspot/share/runtime/safepointMechanism.inline.hpp:67]: Thank you for the explanation.

**Comment by AntonKozlov** [comment]
Thank you for reviews!

/integrate

**Pull request closed without merging**